### PR TITLE
feat: Product Bundle Balance Report

### PR DESCRIPTION
**Pull request by SaiFi0102**
Shows balance qty of product bundles at a given date

### Collapsed
PBB-Collapsed

### Expanded
PBB-Expanded

### CODE REVIEW

File: erpnext/stock/report/product_bundle_balance/product_bundle_balance.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ "report_type": "Script Report", 
+ "roles": [
+  {
+   "role": "Accounts Manager"
+  }, 
+  {
+   "role": "Stock User"
+  }, 
+  {
+   "role": "Warehouse Incharge TS"
```
Comment (nabinhait) [inline erpnext/stock/report/product_bundle_balance/product_bundle_balance.json]: Can you please remove these custom roles?

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/stock/report/product_bundle_balance/product_bundle_balance.json (hunk 1 above)
Comment (SaiFi0102) [reply erpnext/stock/report/product_bundle_balance/product_bundle_balance.json]: Done

**Comment by SaiFi0102** [comment]
`hotfix` or `develop`?

**Comment by rohitwaghchaure** [comment]
@SaiFi0102 Kindly send the new features in the develop branch, only fixes will push on the hotfix branch

File: erpnext/stock/report/product_bundle_balance/product_bundle_balance.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	conditions = " and ".join(conditions)
+	return "and {0}".format(conditions) if conditions else ""
+
+
+def get_sle_conditions(filters):
+	conditions = ""
+	if not filters.get("date"):
+		frappe.throw(_("'Date' is required"))
+
+	conditions += " and sle.posting_date <= '%s'" % frappe.db.escape(filters.get("date"))
```
Comment (deepeshgarg007) [inline erpnext/stock/report/product_bundle_balance/product_bundle_balance.py]: No need to put ```%s``` inside quotes if you are escaping
I am getting this error
image
```suggestion
	conditions += " and sle.posting_date <= %s" % frappe.db.escape(filters.get("date"))
```

File: erpnext/stock/report/product_bundle_balance/product_bundle_balance.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	for d in sle:
+		stock_balance.setdefault(d.item_code, frappe._dict())[d.warehouse] = d
+	return stock_balance
+
+
+def get_stock_ledger_entries(filters, items):
+	if not items:
+		return []
+
+	item_conditions_sql = ' and sle.item_code in ({})' \
```
Comment (deepeshgarg007) [inline erpnext/stock/report/product_bundle_balance/product_bundle_balance.py]: Also the `item_conditions_sql` is not getting formatted properly
Screenshot 2019-05-29 at 7 16 07 PM
```suggestion
	item_conditions_sql = ' and sle.item_code in (%s)' \
```

File: erpnext/stock/report/product_bundle_balance/product_bundle_balance.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		.format(', '.join(['"' + frappe.db.escape(i, percent=False) + '"' for i in items]))
```
Comment (deepeshgarg007) [inline erpnext/stock/report/product_bundle_balance/product_bundle_balance.py]:
```suggestion
		% (', '.join(['%s']*len(items)))
```

File: erpnext/stock/report/product_bundle_balance/product_bundle_balance.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+	return frappe.db.sql("""
+		select
+			sle.item_code, sle.warehouse, sle.qty_after_transaction, sle.company
+		from
+			`tabStock Ledger Entry` sle force index (posting_sort_index)
+		left join `tabStock Ledger Entry` sle2 on
+			sle.item_code = sle2.item_code and sle.warehouse = sle2.warehouse
+			and (sle.posting_date, sle.posting_time, sle.name) < (sle2.posting_date, sle2.posting_time, sle2.name)
+		where sle2.name is null and sle.docstatus < 2 %s %s""" % (item_conditions_sql, conditions), as_dict=1)  # nosec
```
Comment (deepeshgarg007) [inline erpnext/stock/report/product_bundle_balance/product_bundle_balance.py]:
```suggestion
		where sle2.name is null and sle.docstatus < 2 %s %s""" % (item_conditions_sql, conditions), tuple([i for i in items]), as_dict=1)  # nosec nosec
```

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

**Comment by SaiFi0102** [comment]
@deepeshgarg007 Thanks for pointing it out... It's working now

**Review by deepeshgarg007 — APPROVED** [review]

**Pull request merged**